src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts

```diff
@@ -66,6 +69,10 @@ export class NativeLifecycleService extends AbstractLifecycleService {
 			// trigger onWillShutdown events and joining
 			await this.handleWillShutdown(reply.reason);
 
+			// now that all services have stored their data, it's safe to terminate
+			// the remote connection gracefully before synchronously saying we're shut down
+			await this.handleRemoteAgentDisconnect();
```

Comment:
This seems like a strange dependency to me, why can the remote agent service not prolong the shutdown joining the event?

Reply:
In the PR description I mentioned why the two existing points aren't good fits:

> we don't want to do this in
> onWillShutdown because some contributions want to save data to the
> remote workspace, and if shutdown is vetoed it would leave a broken
> state. But onDidShutdown is synchronous and already depended upon by
> several other points, and changing that also felt a bit risky [and would churn a lot]

Reply:
Oh, nevermind., rereading today I see that shutdown can only be vetoed on beforeShutdown, not willShutdown. I can adjust that. I think the risk is if anyone in `onWillShutdown` has to do things like saving state to the remote agent. For example, the edit sessions contributions saves editors and interacts with the filesystem in order to generate its working copies. Maybe we introduce some kind of ordering in the joiners?

Reply:
I would not be opposed to having some kind of priority in the `join` method that would allow contributions to signal whether they have dependencies on other things and thus want early handling vs. late handling. We could simply split up the promise waiters here based on that:

https://github.com/microsoft/vscode/blob/8cb093f3643daf81765d8200afaa602eb03a1bb4/src/vs/workbench/services/lifecycle/electron-sandbox/lifecycleService.ts#L165-L171

Reply:
Added and cleaned up :)